programs/wen-new-standard/src/instructions/mint/add_metadata.rs

```diff
@@ -0,0 +1,70 @@
+use std::str::FromStr;
+
+use anchor_lang::{prelude::*, solana_program::entrypoint::ProgramResult};
+
+use anchor_spl::token_interface::{
+    spl_token_metadata_interface::state::Field, token_metadata_update_field, Mint, Token2022,
+    TokenMetadataUpdateField, TokenMetadataUpdateFieldArgs,
+};
+
+use crate::{errors::MetadataErrors, update_account_lamports_to_minimum_balance2};
+
+#[derive(AnchorDeserialize, AnchorSerialize)]
+pub struct AddMetadataArgs {
+    pub field: String,
+    pub value: String,
+}
+
+#[derive(Accounts)]
+pub struct AddMetadata<'info> {
+    #[account(mut)]
+    pub payer: Signer<'info>,
+    #[account(mut)]
+    pub authority: Signer<'info>,
+    #[account(
+        mut,
+        mint::token_program = token_program,
+    )]
+    pub mint: Box<InterfaceAccount<'info, Mint>>,
+    pub system_program: Program<'info, System>,
+    pub token_program: Program<'info, Token2022>,
+}
+
+impl<'info> AddMetadata<'info> {
+    fn update_token_metadata_field(&self, field: Field, value: String) -> ProgramResult {
+        let cpi_accounts = TokenMetadataUpdateField {
+            token_program_id: self.token_program.to_account_info(),
+            metadata: self.mint.to_account_info(),
+            update_authority: self.authority.to_account_info(),
+        };
+        let cpi_ctx = CpiContext::new(self.token_program.to_account_info(), cpi_accounts);
+        token_metadata_update_field(cpi_ctx, TokenMetadataUpdateFieldArgs { field, value })?;
+        Ok(())
+    }
+}
+
+pub fn handler(ctx: Context<AddMetadata>, args: Vec<AddMetadataArgs>) -> Result<()> {
+    for metadata_arg in args {
+        // validate that the field is not a publickey
+        match Pubkey::from_str(&metadata_arg.field) {
+            Ok(_) => {
+                return Err(MetadataErrors::InvalidField.into());
+            }
+            Err(_) => {
+                ctx.accounts.update_token_metadata_field(
+                    Field::Key(metadata_arg.field),
+                    metadata_arg.value.to_string(),
+                )?;
+            }
+        }
+    }
+
+    // transfer minimum rent to mint account
+    update_account_lamports_to_minimum_balance2(
```

Comment:
is this necessary? isn't metadata space reserved on init?

Reply:
Yeah but every field you add needs to be allocated more space on the mint. Cause the init space will be based on the additional_metadata field in the init instruction.

Reply:
Yes, misunderstood how update CPI worked.